terraform_utils/flatmap.go

```diff
@@ -0,0 +1,333 @@
+	}
+	if countStr == hcl2shim.UnknownVariableValue {
+		return nil, nil
+	}
+
+	count, err := strconv.Atoi(countStr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid count value for %q in state: %s", prefix, err)
+	}
+	if count != len(tys) {
+		return nil, fmt.Errorf("wrong number of values for %q in state: got %d, but need %d", prefix, count, len(tys))
+	}
+
+	var values []interface{}
+	for i, ty := range tys {
+		key := prefix + strconv.Itoa(i)
+		value, err := p.fromFlatmapValue(key, ty)
+		if err != nil {
+			return nil, err
+		}
+		values = append(values, value)
+	}
+	if len(values) == 0 {
+		return nil, nil
+	}
+	return values, nil
+}
+
+func (p *FlatmapParser) fromFlatmapMap(prefix string, ty cty.Type) (map[string]interface{}, error) {
+	// if the container is unknown, there is no count string
+	listName := strings.TrimRight(prefix, ".")
+	if p.attributes[listName] == hcl2shim.UnknownVariableValue {
+		return nil, nil
+	}
+
+	// We actually don't really care about the "count" of a map for our
+	// purposes here, but we do need to check if it _exists_ in order to
+	// recognize the difference between null (not set at all) and empty.
+	strCount, exists := p.attributes[prefix+"%"]
+	if !exists {
+		return nil, nil
+	}
+	if strCount == hcl2shim.UnknownVariableValue {
+		return nil, nil
+	}
+
+	values := make(map[string]interface{})
+	for fullKey := range p.attributes {
+		if !strings.HasPrefix(fullKey, prefix) {
+			continue
+		}
+
+		// The flatmap format doesn't allow us to distinguish between keys
+		// that contain periods and nested objects, so by convention a
+		// map is only ever of primitive type in flatmap, and we just assume
+		// that the remainder of the raw key (dots and all) is the key we
+		// want in the result value.
+		key := fullKey[len(prefix):]
+		if key == "%" {
+			// Ignore the "count" key
+			continue
+		}
+
+		if p.isAttributeIgnored(key, prefix) {
+			continue
+		}
+
+		value, err := p.fromFlatmapValue(fullKey, ty)
+		if err != nil {
+			return nil, err
+		}
+		if p.isValueAllowed(value, prefix) {
+			values[key] = value
+		}
+	}
+	if len(values) == 0 {
+		return nil, nil
+	}
+	return values, nil
+}
+
+func (p *FlatmapParser) fromFlatmapList(prefix string, ty cty.Type) ([]interface{}, error) {
+	// if the container is unknown, there is no count string
+	listName := strings.TrimRight(prefix, ".")
+	if p.attributes[listName] == hcl2shim.UnknownVariableValue {
+		return nil, nil
+	}
+
+	countStr, exists := p.attributes[prefix+"#"]
+	if !exists {
+		return nil, nil
+	}
+	if countStr == hcl2shim.UnknownVariableValue {
+		return nil, nil
+	}
+
+	count, err := strconv.Atoi(countStr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid count value for %q in state: %s", prefix, err)
+	}
+
+	if count == 0 {
+		return nil, nil
+	}
+
+	var values []interface{}
+	for i := 0; i < count; i++ {
+		key := prefix + strconv.Itoa(i)
+		value, err := p.fromFlatmapValue(key, ty)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
add here this code 
```
if value == nil {
	continue
}
if v, ok := value.(map[string]interface{}); ok && len(v) == 0 {
	continue
}
```
sometime providers return array with count=1, but without elements. In this case new code return `null` instead empty array and terraformer return error in HCL parser `error parsing terraform json: unexpected token while parsing list: NULL`

Examples:
kubernetes provider:  deployment.spec.container.resources
aws: elasticache

Reply:
I've changed my `isValueAllowed` function to look for Go's nil and zero values. Then I used it instead of your code snippet, so this should solve the problem.

Reply:
I've tested by importing my ElastiCache infrastructure and it appears to be fine!

Reply:
Thanks! I will test this with same cases.
Maybe you changes fix other issue with Go zero value